staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -758,13 +776,28 @@ func (s *store) ttlOpts(ctx context.Context, ttl int64) ([]clientv3.OpOption, er
 	if ttl == 0 {
 		return nil, nil
 	}
-	// TODO: one lease per ttl key is expensive. Based on current use case, we can have a long window to
-	// put keys within into same lease. We shall benchmark this and optimize the performance.
-	lcr, err := s.client.Lease.Grant(ctx, ttl)
-	if err != nil {
-		return nil, err
+	s.leaseLock.Lock()
+	defer s.leaseLock.Unlock()
+	// If the previous lease grants valid TTL, reuse it; otherwise create
+	// a new one with a little extra time to make it possible to be
+	// reused. We guarantee difference between requested TTL and the
+	// actually used TTL is within the mimimum of
+	// leaseReuseDurationPercent and leaseReuseDurationSeconds.
+	now := time.Now().Unix()
+	remainingTTL := s.prevLeaseExpirationTime - now
+	minReuseDurationSeconds := int64(s.leaseReuseDurationPercent * float64(ttl))
+	if minReuseDurationSeconds > s.leaseReuseDurationSeconds {
+		minReuseDurationSeconds = s.leaseReuseDurationSeconds
+	}
+	if ttl > remainingTTL || ttl+minReuseDurationSeconds < remainingTTL {
+		s.prevLeaseExpirationTime = now + ttl + minReuseDurationSeconds
+		lcr, err := s.client.Lease.Grant(ctx, ttl+minReuseDurationSeconds)
+		if err != nil {
+			return nil, err
+		}
+		s.prevLeaseID = lcr.ID
```

Comment:
Seems reasonable, needs tests

Reply:
Yes, extract the math into a separate method and test that, please.

Reply:
done